### PR TITLE
CMake for reg2grb2

### LINKED ISSUE #263: port reg2grb2 to cmake
Once the global-workflow is done porting to CMake and that work is copied over to coupled-crow, the reg2grb2 utility will need to be ported to CMake.

### PR DESCRIPTION
**Pull request by BrianCurtis-NOAA**
Add a CMakeLists.txt to be able to create reg2grb2.x for coupled-crow after the merge of CMake work from https://github.com/NOAA-EMC/global-workflow/pull/264

Linked issue:
#263 

Waiting on:
Pull #264

### CODE REVIEW

**Comment by JessicaMeixner-NOAA** [comment]
Maybe it's just me, but this seems it could be independent of  #264 and doesn't really need to wait for it?In addition, as this does not delete sorc/reg2grb2.fd/Makefile or update the build script sorc/build_reg2grb2.sh  and is just adding a file, couldn't this just go in now?

**Comment by BrianCurtis-NOAA** [comment]
I did test the file manually putting it into Rahul's CMake work with feature/cmake, and it built the executable.
You are correct that the file can go in now without any issue.
Once the cmake work is brought over to feature/coupled-crow I can make the edits to CMakeLists.txt get it where it needs to be.

**Comment by KateFriedman-NOAA** [comment]
@BrianCurtis-NOAA Checking in...is this draft PR still needed? Thanks!

**Comment by BrianCurtis-NOAA** [comment]
Has reg2grb2 been removed or already ported to cmake elsewhere? I believe I was waiting on the CMake build to be ported into coupled-crow.

**Comment by BrianCurtis-NOAA** [comment]
Yeah missed the cmake port completely. Let me get this added to the sorc CMakeLists.txt.
@WalterKolczynski-NOAA how would I test this is working once I get these last changes in?

**Comment by WalterKolczynski-NOAA** [comment]
As long as it builds and runs, it's fine as far as I'm concerned. I don't have any tests to confirm output for reg2grb2.

**Comment by BrianCurtis-NOAA** [comment]
OK, I'll test on hera and switch this to ready when done.

**Comment by BrianCurtis-NOAA** [comment]
Moving to ready.

**Comment by BrianCurtis-NOAA** [comment]
Please hold for now while I check on something.

**Comment by BrianCurtis-NOAA** [comment]
I left out the find package calls, my reg2grb2 builds OK, but the build on sorc fails with the `enkf_chgres_recenter_nc.fd` directory 
```
[ 19%] Building Fortran object regrid_nemsio.fd/CMakeFiles/regrid_nemsio.x.dir/netcdfio_interface.f90.o
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(5): error #7002: Error in opening the compiled module file.  Check INCLUDE paths.   [MODULE_NCIO]
 use module_ncio
-----^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(53): error #6457: This derived type name has not been declared.   [DATASET]
 type(Dataset) :: indset
------^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(54): error #6457: This derived type name has not been declared.   [DIMENSION]
 type(Dimension) :: ncdim
------^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(319): error #6457: This derived type name has not been declared.   [DATASET]
 type(Dataset) :: refdset
------^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(68): error #6404: This name does not have a type, and must have an explicit type.   [INDSET]
 indset = open_dataset(input_file)
-^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(68): error #6404: This name does not have a type, and must have an explicit type.   [OPEN_DATASET]
 indset = open_dataset(input_file)
----------^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(71): error #6404: This name does not have a type, and must have an explicit type.   [NCDIM]
 ncdim = get_dim(indset, 'grid_xt'); i_input = ncdim%len
-^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(71): error #6404: This name does not have a type, and must have an explicit type.   [GET_DIM]
 ncdim = get_dim(indset, 'grid_xt'); i_input = ncdim%len
---------^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(71): error #6460: This is not a field name that is defined in the encompassing structure.   [LEN]
 ncdim = get_dim(indset, 'grid_xt'); i_input = ncdim%len
-----------------------------------------------------^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(74): error #6404: This name does not have a type, and must have an explicit type.   [GET_IDATE_FROM_TIME_UNITS]
 idate = get_idate_from_time_units(indset)
---------^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(174): error #6632: Keyword arguments are invalid without an explicit interface.   [ERRCODE]
 call read_vardata(indset, 'dzdt', work3d, errcode=iret)
-------------------------------------------^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(192): error #6632: Keyword arguments are invalid without an explicit interface.   [ERRCODE]
 call read_vardata(indset, 'rwmr', work3d, errcode=iret)
-------------------------------------------^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(209): error #6632: Keyword arguments are invalid without an explicit interface.   [ERRCODE]
 call read_vardata(indset, 'icmr', work3d, errcode=iret)
-------------------------------------------^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(226): error #6632: Keyword arguments are invalid without an explicit interface.   [ERRCODE]
 call read_vardata(indset, 'snmr', work3d, errcode=iret)
-------------------------------------------^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(243): error #6632: Keyword arguments are invalid without an explicit interface.   [ERRCODE]
 call read_vardata(indset, 'grle', work3d, errcode=iret)
-------------------------------------------^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(261): error #6632: Keyword arguments are invalid without an explicit interface.   [ERRCODE]
    call read_vardata(indset, 'cld_amt', work3d, errcode=iret)
-------------------------------------------------^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(280): error #6632: Keyword arguments are invalid without an explicit interface.   [ERRCODE]
 call read_vardata(indset, 'dpres', work3d, errcode=iret)
--------------------------------------------^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(286): error #6632: Keyword arguments are invalid without an explicit interface.   [ERRCODE]
 call read_vardata(indset, 'delz', work3d, errcode=iret)
-------------------------------------------^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(324): error #6404: This name does not have a type, and must have an explicit type.   [REFDSET]
 refdset = open_dataset(ref_file)
-^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(324): error #6404: This name does not have a type, and must have an explicit type.   [OPEN_DATASET]
 refdset = open_dataset(ref_file)
-----------^
compilation aborted for /scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90 (code 1)
make[2]: *** [enkf_chgres_recenter_nc.fd/CMakeFiles/enkf_chgres_recenter_nc.x.dir/input_data.f90.o] Error 1
make[1]: *** [enkf_chgres_recenter_nc.fd/CMakeFiles/enkf_chgres_recenter_nc.x.dir/all] Error 2
```
I'm not sure if you want to get that fixed before merging, as reg2grb2 build succeeds.

**Comment by WalterKolczynski-NOAA** [comment]
Are there no changes needed to build_reg2grb2.sh or reg2grb2.fd/Makefile to use cmake?

**Comment by BrianCurtis-NOAA** [comment]
@WalterKolczynski-NOAA OK so the build issue above was actually an issue with hpc-stack that I got taken care of (at least on hera). Make sure that the ncio build from hpc-stack has a module_ncio.mod and not module_fv3gfs_ncio.mod. The script and the makefile are no longer needed. The build completes on Hera.

**Review by WalterKolczynski-NOAA — APPROVED** [review]

**Comment by WalterKolczynski-NOAA** [comment]
@BrianCurtis-NOAA Take a look at the additional changes I made. System now builds using build_all and link_fv3gfs creates the appropriate link to the executable in the exec directory.

**Comment by BrianCurtis-NOAA** [comment]
Looks good to me! Thanks.

File: sorc/build_workflow_utils.sh [hunk 1]
```diff
@@ -5,35 +5,19 @@ set -eux
 readonly UTILS_DIR=$(cd "$(dirname "$($cmd -f -n "${BASH_SOURCE[0]}" )" )" && pwd -P)
 
 # Adapt for global-workflow structure.
-target=${target:-"NULL"}
+source ${UTILS_DIR}/machine-setup.sh > /dev/null 2>&1
```
Comment (aerorahul) [inline sorc/build_workflow_utils.sh:8]: why is this required?
which mahcine is requiring this to be sourced that supports the global workflow?

File: sorc/build_workflow_utils.sh [hunk 2]
```diff
@@ -5,35 +5,19 @@ set -eux
...
 modulefile=${UTILS_DIR}/../modulefiles/workflow_utils.$target
 if [[ -f $modulefile ]]; then
   set +x
-  source ${UTILS_DIR}/machine-setup.sh > /dev/null 2>&1
   source $modulefile
   module list
   set -x
+else
+  echo "FATAL: modulefile $modulefile not found!"
+  exit 1
```
Comment (aerorahul) [inline sorc/build_workflow_utils.sh:17]: This assumes that on every machine there has to be a modulefile.
If the software stack is in path, a modulefile is not required.
Which machine is this failing on?

File: sorc/reg2grb2.fd/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+set(exe_name reg2grb2.x)
+add_executable(${exe_name} ${fortran_src})
+target_link_libraries(
+  ${exe_name}
+  landsfcutil::landsfcutil_d
+  ip::ip_d
+  sp::sp_d
+  bacio::bacio_4
+  w3nco::w3nco_4
```
Comment (aerorahul) [inline sorc/reg2grb2.fd/CMakeLists.txt:20]: has anyone verified that all of these are actually used in this program?

**Review by aerorahul — COMMENTED** [review]
please see comments.

File: sorc/reg2grb2.fd/CMakeLists.txt (hunk 3 above)
Comment (BrianCurtis-NOAA) [reply sorc/reg2grb2.fd/CMakeLists.txt:20]: I went through the makefile and compared which ones were being linked. I didn't go through the code line by line.

File: sorc/build_workflow_utils.sh (hunk 1 above)
Comment (WalterKolczynski-NOAA) [reply sorc/build_workflow_utils.sh:8]: When I tried running build_workflow_utils stand-alone, $target isn't defined until it is set by machine-setup, so it was using NULL, then quietly not loading a modulefile because there wasn't one for NULL, then cmake would fail.

File: sorc/build_workflow_utils.sh (hunk 2 above)
Comment (WalterKolczynski-NOAA) [reply sorc/build_workflow_utils.sh:17]: The need for this came to me based on the above comment. In the context of global-workflow, I don't know that manually loading a module beforehand is a use case to support, and I'd prefer the error to specify the actual problem rather than cmake just failing.

File: sorc/reg2grb2.fd/CMakeLists.txt (hunk 3 above)
Comment (WalterKolczynski-NOAA) [reply sorc/reg2grb2.fd/CMakeLists.txt:20]: That's a question for @BrianCurtis-NOAA, but landsfcutil is a new dependency for this PR, so I assume yes for that at least.

Comment (BrianCurtis-NOAA) [reply sorc/reg2grb2.fd/CMakeLists.txt:20]: @aerorahul the old build_reg2grb2.sh has landsfcutil in a LIBS export which was then used in the makefile that was used to build reg2grb2. I don't see any subroutines that seem to be using any landsfcutil subroutines, but if needed I will spend more time on going line by line to see if i missed something. Let me know.

**Comment by WalterKolczynski-NOAA** [comment]
@aerorahul Unless you want me to hold this, I'd like to get it merged.

**Comment by aerorahul** [comment]
> @aerorahul Unless you want me to hold this, I'd like to get it merged.

Please go ahead.
